internal/plugin/connectors/http/generic/oauthprotocol/protocol.go

```diff
@@ -0,0 +1,214 @@
+	oauth_timestamp        = "oauth_timestamp"
+
+	oauth_version_number        = "1.0"
+	oauth_signature_method_type = "HMAC-SHA1"
+	nonce_length                = 16
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyz" +
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+	"0123456789"
+
+func stringWithCharset(length int, charset string) string {
+	var seededRand *rand.Rand = rand.New(
+		rand.NewSource(time.Now().UnixNano()))
+
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[seededRand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func generateNonce(length int) string {
+	return stringWithCharset(length, charset)
+}
+
+func checkRequiredOAuthParams(params map[string]string) error {
+	var requiredParams = []string{
+		oauth_consumer_key,
+		oauth_consumer_secret,
+		oauth_token,
+		oauth_token_secret,
+	}
+
+	for _, requiredParam := range requiredParams {
+		if param := params[requiredParam]; len(param) <= 0 {
+			return fmt.Errorf("required OAuth1 parameter '%s' not found", requiredParam)
+		}
+	}
+	return nil
+}
+
+// ExtractOAuthParams extracts the required OAuth1 parameters from the config
+// and assigns the other required paramters
+// If a required parameter is missing from the config throws an error
+func extractOAuthParams(params map[string]string) OAuth1 {
+	var oauth OAuth1
+
+	oauth.ConsumerKey = params[oauth_consumer_key]
+	oauth.ConsumerSecret = params[oauth_consumer_secret]
+	oauth.Token = params[oauth_token]
+	oauth.TokenSecret = params[oauth_token_secret]
+	oauth.TimeStamp = strconv.Itoa(int(time.Now().Unix()))
+	oauth.Version = oauth_version_number
+	oauth.SignatureMethod = oauth_signature_method_type
+	oauth.Nonce = generateNonce(nonce_length)
+
+	return oauth
+}
+
+func encodeURI(stringToEncode string) string {
+	return strings.ReplaceAll(url.QueryEscape(stringToEncode), "+", "%20")
+}
+
+// Create a Parameter String from the required OAuth params and query & body params
+// https://tools.ietf.org/html/rfc5849#section-3.5
+func generateParameterString(oauth1 OAuth1, r *gohttp.Request) string {
+	// create map of all parameters
+	buf, _ := ioutil.ReadAll(r.Body)
+	originalBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	copyBody := ioutil.NopCloser(bytes.NewBuffer(buf))
+	r.Body = originalBody
+
+	var parameterString bytes.Buffer
+	// create map of all parameters
+	paramMap := make(map[string][]string, 0)
+	// get query and body params and URL encode the keys(values URL encoded later)
+	r.ParseForm()
+	for paramKey, paramValues := range r.Form {
+		encodedKey := encodeURI(paramKey)
+		paramMap[encodedKey] = paramValues
+	}
+	r.Body = copyBody
+
+	// these keys are already URL safe
+	paramMap[oauth_consumer_key] = []string{oauth1.ConsumerKey}
+	paramMap[oauth_nonce] = []string{oauth1.Nonce}
+	paramMap[oauth_signature_method] = []string{oauth1.SignatureMethod}
+	paramMap[oauth_timestamp] = []string{oauth1.TimeStamp}
+	paramMap[oauth_token] = []string{oauth1.Token}
+	paramMap[oauth_version] = []string{oauth1.Version}
```

Comment:
Ditto not sure why this needs to be an array of strings

Reply:
These need to be a string array to match the `map[string][]string` that we get from `r.Form`. It made a lot more sense logically to lump both of these parameter maps together, and use the same logic that follows to extract, and append the values together to make the string